File: instrumentation/akka-http-2.13_10.2.0/src/main/scala/com/agent/instrumentation/akka/http/FutureWrapper.scala
package com.agent.instrumentation.akka.http

import com.newrelic.agent.bridge.AgentBridge
import com.newrelic.api.agent.NewRelic
import com.newrelic.api.agent.weaver.Weaver

import scala.concurrent.{CanAwait, ExecutionContext, Future}
import scala.concurrent.duration.Duration
import scala.reflect.ClassTag
import scala.util.Try

class FutureWrapper[T](val original: Future[T]) extends Future[T] {

  override def onComplete[U](f: Try[T] => U)(implicit executor: ExecutionContext): Unit = {
    try {
      val wrappedFunction: Try[T] => U = new Function1Wrapper(f, NewRelic.getAgent.getTransaction.getToken)
      original.onComplete(wrappedFunction)
    } catch {
      case t: Throwable => {
        AgentBridge.instrumentation.noticeInstrumentationError(t, Weaver.getImplementationTitle)
        original.onComplete(f)
      }
    }
  }

  override def isCompleted: Boolean = original.isCompleted

  override def value: Option[Try[T]] = original.value

  override def failed: Future[Throwable] = original.failed

  override def transform[S](f: Try[T] => Try[S])(implicit executor: ExecutionContext): Future[S] = original.transform(f)

  override def transformWith[S](f: Try[T] => Future[S])(implicit executor: ExecutionContext): Future[S] = original.transformWith(f)

  override def ready(atMost: Duration)(implicit permit: CanAwait): this.type = {
    original.ready(atMost)
    this
  }

  override def result(atMost: Duration)(implicit permit: CanAwait): T = original.result(atMost)

  override def foreach[U](f: T => U)(implicit executor: ExecutionContext): Unit = original.foreach(f)

  override def transform[S](s: T => S, f: Throwable => Throwable)(implicit executor: ExecutionContext): Future[S] = original.transform(s, f)

  override def map[S](f: T => S)(implicit executor: ExecutionContext): Future[S] = original.map(f)

  override def flatMap[S](f: T => Future[S])(implicit executor: ExecutionContext): Future[S] = original.flatMap(f)

  override def flatten[S](implicit ev: T <:< Future[S]): Future[S] = original.flatten

  override def filter(p: T => Boolean)(implicit executor: ExecutionContext): Future[T] = original.filter(p)

  override def collect[S](pf: PartialFunction[T, S])(implicit executor: ExecutionContext): Future[S] = original.collect(pf)

  override def recover[U >: T](pf: PartialFunction[Throwable, U])(implicit executor: ExecutionContext): Future[U] = original.recover(pf)

  override def recoverWith[U >: T](pf: PartialFunction[Throwable, Future[U]])(implicit executor: ExecutionContext): Future[U] = original.recoverWith(pf)

  override def zip[U](that: Future[U]): Future[(T, U)] = original.zip(that)

  override def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = original.zipWith(that)(f)

  override def fallbackTo[U >: T](that: Future[U]): Future[U] = original.fallbackTo(that)

  override def mapTo[S](implicit tag: ClassTag[S]): Future[S] = original.mapTo

  override def andThen[U](pf: PartialFunction[Try[T], U])(implicit executor: ExecutionContext): Future[T] = original.andThen(pf)

}

File: instrumentation/apache-pekko-http-3_1/src/main/scala/com/agent/instrumentation/org.apache.pekko/http/FutureWrapper.java
/*
 *
 *  * Copyright 2024 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package com.agent.instrumentation.org.apache.pekko.http;

import com.newrelic.agent.bridge.AgentBridge;
import com.newrelic.api.agent.NewRelic;
import com.newrelic.api.agent.weaver.Weaver;
import scala.$less$colon$less;
import scala.Function1;
import scala.Function2;
import scala.Option;
import scala.PartialFunction;
import scala.Tuple2;
import scala.concurrent.Awaitable;
import scala.concurrent.CanAwait;
import scala.concurrent.ExecutionContext;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;
import scala.reflect.ClassTag;
import scala.util.Try;

import java.util.concurrent.TimeoutException;

public class FutureWrapper<T> implements Future<T> {

    private final Future<T> original;

    public FutureWrapper(Future<T> original) {
        this.original = original;
    }

    @Override
    public <U> void onComplete(Function1<Try<T>, U> f, ExecutionContext executor) {
        try {
            f = new Function1Wrapper<>(f, NewRelic.getAgent().getTransaction().getToken());
        } catch (Throwable t) {
            AgentBridge.instrumentation.noticeInstrumentationError(t, Weaver.getImplementationTitle());
        }
        original.onComplete(f, executor);
    }

    @Override
    public boolean isCompleted() {
        return original.isCompleted();
    }

    @Override
    public Option<Try<T>> value() {
        return original.value();
    }

    @Override
    public Future<Throwable> failed() {
        return original.failed();
    }

    @Override
    public <U> void foreach(Function1<T, U> f, ExecutionContext executor) {
        original.foreach(f, executor);
    }

    @Override
    public <S> Future<S> transform(Function1<T, S> s, Function1<Throwable, Throwable> f, ExecutionContext executor) {
        return original.transform(s, f, executor);
    }

    @Override
    public <S> Future<S> transform(Function1<Try<T>, Try<S>> f, ExecutionContext executor) {
        return original.transform(f, executor);
    }

    @Override
    public <S> Future<S> transformWith(Function1<Try<T>, Future<S>> f, ExecutionContext executor) {
        return original.transformWith(f, executor);
    }

    @Override
    public <S> Future<S> map(Function1<T, S> f, ExecutionContext executor) {
        return original.map(f, executor);
    }

    @Override
    public <S> Future<S> flatMap(Function1<T, Future<S>> f, ExecutionContext executor) {
        return original.flatMap(f, executor);
    }

    @Override
    public <S> Future<S> flatten($less$colon$less<T, Future<S>> ev) {
        return original.flatten(ev);
    }

    @Override
    public Future<T> filter(Function1<T, Object> p, ExecutionContext executor) {
        return original.filter(p, executor);
    }

    @Override
    public Future<T> withFilter(Function1<T, Object> p, ExecutionContext executor) {
        return original.withFilter(p, executor);
    }

    @Override
    public <S> Future<S> collect(PartialFunction<T, S> pf, ExecutionContext executor) {
        return original.collect(pf, executor);
    }

    @Override
    public <U> Future<U> recover(PartialFunction<Throwable, U> pf, ExecutionContext executor) {
        return original.recover(pf, executor);
    }

    @Override
    public <U> Future<U> recoverWith(PartialFunction<Throwable, Future<U>> pf, ExecutionContext executor) {
        return original.recoverWith(pf, executor);
    }

    @Override
    public <U> Future<Tuple2<T, U>> zip(Future<U> that) {
        return original.zip(that);
    }

    @Override
    public <U, R> Future<R> zipWith(Future<U> that, Function2<T, U, R> f, ExecutionContext executor) {
        return original.zipWith(that, f, executor);
    }

    @Override
    public <U> Future<U> fallbackTo(Future<U> that) {
        return original.fallbackTo(that);
    }

    @Override
    public <S> Future<S> mapTo(ClassTag<S> tag) {
        return original.mapTo(tag);
    }

    @Override
    public <U> Future<T> andThen(PartialFunction<Try<T>, U> pf, ExecutionContext executor) {
        return original.andThen(pf, executor);
    }

    @Override
    public Awaitable<T> ready(Duration atMost, CanAwait permit) throws InterruptedException, TimeoutException {
        return original.ready(atMost, permit);
    }

    @Override
    public T result(Duration atMost, CanAwait permit) throws TimeoutException, InterruptedException {
        return original.result(atMost, permit);
    }
}
